src/transformStateWithClones.js

```diff
@@ -6,8 +6,46 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const allStates = [];
+  const realState = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(realState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(realState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(realState);
+        break;
+    }
+
+    allStates.push(Object.assign({}, realState));
```

Comment:
According to the checklist, it is better to push a copy of the object to the array at the end of the loop cycle, but outside of the `switch` block. However, this is already being done correctly here, so no changes are needed.